thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/v2/AggregationResource.java

```diff
@@ -0,0 +1,263 @@
+    if (!StringUtils.isBlank(filterString)) {
+      filters = ThirdEyeUtils.convertToMultiMap(filterString);
+    }
+
+    LOG.info("Requesting {} metrics with {} filters from {} time ranges",
+        metricIds.size(), filters.size(), ranges.size());
+
+    // request data
+    Map<MetricSlice, Future<DataFrame>> requests = new HashMap<>();
+    for (String id : metricIds) {
+      for (Interval range : ranges) {
+        long metricId = Long.valueOf(id);
+        long start = range.getStartMillis();
+        long end = range.getEndMillis();
+        MetricSlice slice = MetricSlice.from(metricId, start, end, filters);
+
+        requests.put(slice, fetchMetricAggregationAsync(slice));
+      }
+    }
+
+    // collect results
+    Map<String, Map<String, Map<String, Map<String, Double>>>> results = new HashMap<>();
+    for (Interval range : ranges) {
+      Map<String, Map<String, Map<String, Double>>> rangeResults = new HashMap<>();
+
+      for (MetricSlice slice : filterSlices(requests.keySet(), range)) {
+        String mid = String.valueOf(slice.getMetricId());
+        DataFrame df = requests.get(slice).get(TIMEOUT, TimeUnit.MILLISECONDS);
+
+        if (rollupCount != null) {
+          df = rollupTail(df, rollupCount);
+        }
+
+        rangeResults.put(mid, dataframeToMap(df));
+      }
+
+      String rid = String.format("%s:%s", range.getStartMillis(), range.getEndMillis());
+      results.put(rid, rangeResults);
+    }
+
+    return results;
+  }
+
+  /**
+   * Helper to roll up dimension values to a fixed count per dimension name. The rolled up column
+   * carries the dimension value {@code COL_DIMENSION_VALUE_OTHER}.
+   *
+   * @param df
+   * @param numValues
+   * @return
+   */
+  private static DataFrame rollupTail(DataFrame df, int numValues) {
+    if (df.isEmpty()) {
+      return df;
+    }
+
+    List<DataFrame> results = new ArrayList<>();
+    for (String dimName : df.getStrings(COL_DIMENSION_NAME).unique().values()) {
+      DataFrame dfValues = df.filterEquals(COL_DIMENSION_NAME, dimName).dropNull().sortedBy(COL_VALUE).reverse();
+
+      DataFrame dfPassthru = dfValues.sliceTo(numValues - 1);
+      DataFrame dfRollup = dfValues.sliceFrom(numValues - 1);
+
+      if (!dfRollup.isEmpty()) {
+        double sum = dfRollup.getDoubles(COL_VALUE).sum().fillNull().doubleValue();
+        DataFrame dfAppend = DataFrame
+            .builder(COL_DIMENSION_NAME, COL_DIMENSION_VALUE, COL_VALUE)
+            .append(dimName, COL_DIMENSION_VALUE_OTHER, sum).build();
+        dfPassthru = dfPassthru.append(dfAppend);
+      }
+
+      results.add(dfPassthru);
+    }
+
+    DataFrame first = results.get(0);
+    results.remove(0);
+
+    return first.append(results);
+  }
+
+  /**
+   * Helper to convert dimension aggregation dataframe to a map of values
+   *
+   * @param df
+   * @return
+   */
+  private static Map<String, Map<String, Double>> dataframeToMap(DataFrame df) {
```

Comment:
Add comment: "Map (keyed by dimension name) of maps (keyed by dimension value) of values".

Reply:
fixed